internal/bgp/bgpfrr/frr.go

```diff
@@ -0,0 +1,150 @@
+// the 'frrState' struct.
+func sessionName(srcAddr string, myAsn uint32, addr string, asn uint32) string {
+	return fmt.Sprintf("%d@%s-%d@%s", asn, addr, myAsn, srcAddr)
+}
+
+func (s *Session) Set(advs ...*bgp.Advertisement) error {
+	newAdvs := map[string]*bgp.Advertisement{}
+	for _, adv := range advs {
+		err := bgp.Validate(adv)
+		if err != nil {
+			return err
+		}
+		newAdvs[adv.Prefix.String()] = adv
+	}
+	s.advertised = newAdvs
+	state.regenerate()
+
+	err := configFRR()
+
+	return err
+}
+
+// Close() shuts down the BGP session.
+func (s *Session) Close() error {
+	state.deleteSession(sessionName(s.srcAddr.String(), s.asn, s.addr, s.peerASN))
+	state.regenerate()
+
+	err := configFRR()
+
+	return err
+}
+
+// New() creates a BGP session using the given session parameters.
+//
+// The session will immediately try to connect and synchronize its
+// local state with the peer.
+func New(l log.Logger, addr string, srcAddr net.IP, asn uint32, routerID net.IP, peerASN uint32, holdTime time.Duration, password string, myNode string) (*Session, error) {
+	session := &Session{
+		asn:        asn,
+		routerID:   routerID.To4(),
+		myNode:     myNode,
+		addr:       addr,
+		srcAddr:    srcAddr,
+		peerASN:    peerASN,
+		holdTime:   holdTime,
+		logger:     log.With(l, "peer", addr, "localASN", asn, "peerASN", peerASN),
+		password:   password,
+		advertised: map[string]*bgp.Advertisement{},
+	}
+
+	state.addSession(session)
+	state.regenerate()
+
+	err := configFRR()
+
+	return session, err
+}
+
+func (s *frrState) addSession(session *Session) {
+	s.sessions[sessionName(session.srcAddr.String(), session.asn, session.addr, session.peerASN)] = session
+}
+
+func (s *frrState) deleteSession(name string) {
+	delete(s.sessions, name)
+}
+
+func (s *frrState) regenerate() {
+	config := frrConfig{
+		Hostname: "", // TODO.
+		Loglevel: "", // TODO.
+		Routers:  make(map[string]*routerConfig),
+	}
+
+	for _, session := range s.sessions {
+		var router *routerConfig
+		var neighbor *neighborConfig
+		var exist bool
+
+		routerName := routerName(session.srcAddr.String(), session.asn)
+		if router, exist = config.Routers[routerName]; !exist {
+			router = &routerConfig{
+				Asn:       session.asn,
+				RouterId:  session.srcAddr.String(),
+				Neighbors: make(map[string]*neighborConfig),
+			}
+			config.Routers[routerName] = router
+		}
+
+		neighborName := neighborName(session.addr, session.peerASN)
+		if neighbor, exist = router.Neighbors[neighborName]; !exist {
+			neighbor = &neighborConfig{
+				Asn:            session.peerASN,
+				Addr:           session.addr,
+				Advertisements: make(map[string]*advertisementConfig),
+			}
+			router.Neighbors[neighborName] = neighbor
+		}
+
+		for _, adv := range session.advertised {
+			if _, exist := neighbor.Advertisements[adv.Prefix.String()]; !exist {
+				neighbor.Advertisements[adv.Prefix.String()] = &advertisementConfig{
+					Version: "ipv4",
+					Prefix:  adv.Prefix.String(),
```

Comment:
do u need to look at prefix to findout the AF instead of hardcoding IPv4 ?

Reply:
+1

Reply:
Yeah that was a hack as we don't support IPv6 but I may as well fix it now. Thanks for catching it.